flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java

```diff
@@ -200,10 +226,45 @@ protected void onReleaseTaskManager(ResourceCounter previouslyFulfilledRequireme
 
     @VisibleForTesting
     void newSlotsAreAvailable(Collection<? extends PhysicalSlot> newSlots) {
+        if (!globalSlotsViewable) {
+            final Collection<RequestSlotMatchingStrategy.RequestSlotMatch> requestSlotMatches =
+                    requestSlotMatchingStrategy.matchRequestsAndSlots(
+                            newSlots, pendingRequests.values());
+            reserveMatchedFreeSlots(requestSlotMatches);
+            fulfillMatchedSlots(requestSlotMatches);
+            return;
+        }
+
+        receivedSlots.addAll(newSlots);
+        if (receivedSlots.size() < pendingRequests.size()) {
+            return;
+        }
         final Collection<RequestSlotMatchingStrategy.RequestSlotMatch> requestSlotMatches =
                 requestSlotMatchingStrategy.matchRequestsAndSlots(
                         newSlots, pendingRequests.values());
+        if (requestSlotMatches.size() == pendingRequests.size()) {
+            reserveMatchedFreeSlots(requestSlotMatches);
+            fulfillMatchedSlots(requestSlotMatches);
+        }
+        receivedSlots.clear();
```

Comment:
Should we only clear the pending slots after all the requirements are fulfilled?

Reply:
The `receivedSlots` will be cleared no matter whether we got enough matching slots.

Reply:
IIUC, Batched resource requests are sent to RM, and generally, all slots on each TM are provided to the `DeclarativeSlotPoolBridge` in a batch, with each call triggered `newAvailableSlots` once. So we need to cache it here, right ?  I'd appreciate it with your  opinion or confirmation .